### DISCUSSION TITLE
[RFC] Environments overhaul

### DISCUSSION THREAD

**Discussion by anthonyshew** [typehero/typehero#500] [General]
# Abstract
Typehero needs to be able to conduct database migrations. Currently, we're working with [PlanetScale](https://planetscale.com/) and [Vercel](https://vercel.com) so it should be possible for us to pave a path where we can always have Vercel previews with an associated PlanetScale branch that has working schema with the application preview.

This has been solved thus far with two separate Vercel projects and handrolling database migration procedures. At the moment, we have no systematic way to handle migrations. Everything is done ad-hoc, as needed. This has been giving us problems since...well, it gets messed up just about every time, it seems, one way or another. 😄

# Problems we need to solve
## Production stability
Typehero currently is only a waitlist on production - but has broken a handful of times already in its first week of going live. We need to gain a bit more stability so we can have confidence in our ships.

## Database migrations
How do we want to conduct database migrations? How do we bring these migrations from branches to production?

## Multiple Vercel projects
At the moment, we have two projects on Vercel...for one project. How do we turn things into one project, shipping to production but still only showing the waitlist to the public?

## Multiple PlanetScale databases
We also currently have multiple PlanetScale databases. I haven't quite understood why this is the case but it's unnecessary given PlanetScale's excellent branching functionality.

# Goals
1. Stability - Working waitlist in production
2. Release management - Only waitlist and home page available to public in production
3. DX - Operational Vercel previews (even with database migration proposals)
4. DX - Consistent database migration procedure
5. Security - Never reveal production database credentials to open source contributors

# Non-goals
- Accessing non-public routes by staff and admin
   - This can be done in follow-up work

# Proposed Solution

We'll use one Vercel project and one PlanetScale database and use branches according to the intentions of the various technologies. Additionally, we will automate database migrations using the respective [PlanetScale](https://planetscale.com/features/cli) and [Vercel](https://vercel.com/docs/cli) CLIs in GitHub Actions.

> Notably, there are several moving parts here - which isn't how I prefer to handle migrations. However, the concerns and goals described above are coupled enough to justify making these changes as one initiative. Additionally, the project is in a low-risk, high-velocity state given that there is strictly a waitlist at this time.

## One Vercel project
We will ship to production under a flag that protects the entire site (except for `/` and `/waitlist`) defined in Vercel Edge Config. Since Edge Config has sub-millisecond read times, we won't impact load times for the public-accessible pages. Additionally, we will create two Edge Configs, one for production and one for previews/development, so we can test things on previews before we release. We'll need to make a few code changes to read from Edge Configs in `middleware.ts` to enable this behavior.

Once we have our feature flagging in place, we'll move on the workflow for database migrations.

## Database migration workflow

By default, all Vercel previews will be linked to a persistent PlanetScale branch. We'll do this so we don't have to constantly make new branches for the same schema (and stay within PlanetScale branching usage limits).

### Getting a preview of changes
1. Developer writes code, introducing a change to the schema defined in `schema.prisma`
2. On pull request, a GitHub Action kicks off that does the following:
   a) Install PlanetScale CLI and Vercel CLI
   b) [Create a database branch](https://planetscale.com/docs/reference/branch)
   c) Push Prisma schema and seed database
   d) [Assign credentials for branch to Vercel branch](https://vercel.com/docs/cli/env#extended-usage)

> Note that this workflow ONLY runs when `schema.prisma` contains changes. This is achievable using [GHA path filters](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore).

We now have a Vercel preview that runs against our new database schema. 🥳

### Getting to production

At this point, we will validate that we like our changes and need to deploy to production. To do so a project maintainer needs to do the following:
1. Review code. Note that database migrations **must always allow the currently running application to be compatible with the incoming schema**. This is because the incoming schema will go out before the application redeploys. ([Theory discussion here.](https://github.com/planetscale/discussion/discussions/175#discussioncomment-5604976))
2. Visit the PlanetScale UI (or CLI) to [create a Deploy Request](https://planetscale.com/docs/onboarding/branching-and-deploy-requests#create-a-deploy-request). If all is well, deploy schema changes.
3. Visit the pull request and merge code changes.
4. Vercel deploys the application to production.

Shipped! 🥳

> Notice that we never need to show production credentials to contributors through this process, satisfying [Goal #5](#goals). I *think* this process doesn't even real preview credentials to contributors, either.

### Cleanup
Several cleanup actions need to be taken:
1. PlanetScale may delete merged branches automatically; I forget. If not, we will need to close old branches to stay within usage limits.
2. Remove the old connection string from the Vercel branch.

Both of these actions can be taken in a GHA action that runs post-merge and a pull request closes.

## Possible failure modes
- A race condition exists where the Vercel preview will try to build before PlanetScale has created a database branch and assigned it to the Vercel branch environment variables. In this case, we can kick off another Vercel build and the build will be successful. (There may be a more graceful way to handle this but this should work for now.)
- PlanetScale has branch usage limitations that we will want to be conscious of. If there are many branches that are making many database schema changes, we will attempt to create too many PlanetScale branches, exceeding our limits. In this situation, the GitHub action will fail. It should be fairly clear what the failure mode is for this case. (I'd also argue that we have larger problems with project management if we run into this situation/limitation with any serious consistency.)
- If an unexpectedly broken migration makes it to production, PlanetScale has [Schema Reverts](https://planetscale.com/features/revert) and Vercel has [Instant Rollback](https://vercel.com/changelog/instant-rollback-is-now-available-to-revert-deployments) so we can revert to a successful application state quickly.

# Alternatives Considered

## Continue manually rolling - but with documentation
Part of the reason we are seeing instability as we work towards production is that we have to relearn the process to make a migration every time we would like to make one.

If we create a runbook so we can follow a happy path every time, we can still choose to handle migrations by hand. I would say this is the path to take if we are uncomfortable with automating the process and believe it to be error prone. (Hopefully, the Proposed Solution sounds solid enough that we don't need to fall back to here.)

If we were to choose this alternative, I would still recommend that we merge to one Vercel project and one PlanetScale database. I won't write out the entire happy path here but can do so if we start to lean towards this solution.

---
Happy to hear any feedback! Please add any discussion or questions you may have here.

Once we have a solution that we are happy with we can put together a PR to put our changes into place.

---

# Changelog
`type NotYet = typeof null`

**Comment by jhechtf** [comment]
Re: Feature Flags with Edge Config, there was some work I did waaaay back to setup https://happykit.dev/ feature flags. It worked pretty well and I managed to get some basic by-user feature flag roll outs. 

There are also prisma migrations, but I think @basokant had more raw knowledge about it than anyone else. IIRC (and Ben tell me if I'm wrong) Prisma migrations are absolutely fine, it just makes sense for us to start doing migrations once the fever of development subsides a bit so there aren't rampant schema changes.

**Reply by basokant** [reply]
I haven't looked into how to get prisma migrations working (especially with how planetscale is kind of responsible for managing schema changes if we're using their service). Ran into this article which should help us https://planetscale.com/docs/prisma/automatic-prisma-migrations

**Reply by jhechtf** [reply]
hmm interesting, there's definitely FAFO to deal with about this.

**Comment by bautistaaa** [comment]
yeah i think at least the  first actionable item is use a branch for staging

**Comment by Hacksore** [comment]
@anthonyshew does every feature branch end up with a 1:1 planetscale branch?
